### PR TITLE
[FIX] im_livechat: remove accent of cookie

### PR DESCRIPTION
**Pull request by nim-odoo**
- Install `website_livechat`
- Publish the demo livechat
- Set an accent to the operator name, e.g. 'Bòb'
- Visit the website with Safari and start a chat session
- Close the window
- Refresh the page

Error in browser console that sometime appear as a traceback.

It happens because Safari refuses to send any cookie containing
non-ASCII characters [1]. Actually, Safari strips the value before the
first non-ASCII character, making the JSON string corrupted.

Having such special characters in cookies is uncommon, so we only clean
the necessary string. Moreover, it's better to avoid our tools modifying
all cookies set without being aware of it.

[1] https://stackoverflow.com/questions/1969232/what-are-allowed-characters-in-cookies/1969339#1969339

opw-2273293

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by nim-odoo** [comment]
We might need a better solution for non-latin characters. On the other hand, I don't want to introduce an overkill solution for this.

**Review by nle-odoo — APPROVED** [review]
I think this reminded me of this issue https://github.com/odoo/odoo/pull/50554 but it seems different


I'm not sure the change would do anything in python side, since by default the output of json.dumps should have `ensure_ascii` (with special characters encoded in `\uHHHH` syntax as allowed by JSON https://www.json.org/):

https://docs.python.org/3/library/json.html#json.dumps

```
>>> json.dumps('école')
'"\\u00e9cole"'
```

as for JS side, an alternative would be to encode special character as done in this link (manually since it's not done by default in JSON.stringify): https://stackoverflow.com/a/31652607

**Comment by nim-odoo** [comment]
Seems working in Safari, thanks

File: addons/im_livechat/static/src/js/im_livechat.js [hunk 1]
```diff
@@ -247,7 +247,15 @@ var LivechatButton = Widget.extend({
...
 
-                    utils.set_cookie('im_livechat_session', JSON.stringify(self._livechat.toData()), 60*60);
+                    // Safari doesn't accept cookies with non-ASCII characters, so we encode the
+                    // string.
+                    // Note that Python does this encoding automatically.
+                    // Taken from https://stackoverflow.com/a/31652607
+                    var im_livechat_session = JSON.stringify(self._livechat.toData());
+                    im_livechat_session  = im_livechat_session.replace(/[\u007F-\uFFFF]/g, function(chr) {
+                        return "\\u" + ("0000" + chr.charCodeAt(0).toString(16)).substr(-4)
+                    })
```
Comment (alexkuhn) [inline addons/im_livechat/static/src/js/im_livechat.js:257]:
```suggestion
                    let im_livechat_session = JSON.stringify(self._livechat.toData());
                    im_livechat_session = im_livechat_session.replace(/[\u007F-\uFFFF]/g, chr =>
                        `\\u${`0000${chr.charCodeAt(0).toString(16)}`.substr(-4)}`
                    );
```

**Comment by nim-odoo** [comment]
That's very weird... I can't figure out why I sometimes get the following:
```
"folded":false,"id":15,"message_unread_counter":0,"operator_pid":[3,"YourCompany Mitchell Admin \u00f2 \u6f22\u8a9e"],"name":"Website Visitor #16 Mitchell Admin \u00f2 \u6f22\u8a9e","uuid":"ad3adefd-619a-47ce-9797-9819e203395a"
```

Which of course cannot be loaded by JSON... Even with the last version of the patch (which only escapes the names). I guess I'm going back to the `unaccent` solution, which seems safer.

**Comment by nle-odoo** [comment]
with current version of PR I'm not able to reproduce the issue

with the previous patch, it seemed to work when I was testing only on one side (with the given scenario) but it broke if the operator replied

with the current one trying several use case (of closing / refreshing / sending message with operator, with public user) I was not able to have an error

and for `"folded":false,"id":15,"message_unread_counter":0,"operator_pid":[3,"YourCompany Mitchell Admin \u00f2 \u6f22\u8a9e"],"name":"Website Visitor #16 Mitchell Admin \u00f2 \u6f22\u8a9e","uuid":"ad3adefd-619a-47ce-9797-9819e203395a"` without curly braces I was not able to get it

**Comment by nim-odoo** [comment]
> with current version of PR I'm not able to reproduce the issue
> 
> with the previous patch, it seemed to work when I was testing only on one side (with the given scenario) but it broke if the operator replied
> 
> with the current one trying several use case (of closing / refreshing / sending message with operator, with public user) I was not able to have an error
> 
> and for `"folded":false,"id":15,"message_unread_counter":0,"operator_pid":[3,"YourCompany Mitchell Admin \u00f2 \u6f22\u8a9e"],"name":"Website Visitor #16 Mitchell Admin \u00f2 \u6f22\u8a9e","uuid":"ad3adefd-619a-47ce-9797-9819e203395a"` without curly braces I was not able to get it

To reproduce, as the visitor go to the website and click on the "Livechat Sessions" menu.

**Comment by nim-odoo** [comment]
Note to myself:
```
    decode_server_cookie: function (val) {
        if (!val.test(/\\(?!u)/)) { // if no \ (excepted \u)
            return val;  // not encoded
        }
```

**Pull request closed without merging**